src/gc/objecthandle.cpp

```diff
@@ -278,39 +275,11 @@ void CALLBACK PinObject(_UNCHECKED_OBJECTREF *pObjRef, uintptr_t *pExtraInfo, ui
     promote_func* callback = (promote_func*) lp2;
     callback(pRef, (ScanContext *)lp1, GC_CALL_PINNED);
 
-#ifndef FEATURE_REDHAWK
     Object * pPinnedObj = *pRef;
-
-    if (!HndIsNullOrDestroyedHandle(pPinnedObj) && pPinnedObj->GetGCSafeMethodTable() == g_pOverlappedDataClass)
+    if (!HndIsNullOrDestroyedHandle(pPinnedObj))
     {
-        // reporting the pinned user objects
-        OverlappedDataObject *pOverlapped = (OverlappedDataObject *)pPinnedObj;
-        if (pOverlapped->m_userObject != NULL)
-        {
-            //callback(OBJECTREF_TO_UNCHECKED_OBJECTREF(pOverlapped->m_userObject), (ScanContext *)lp1, GC_CALL_PINNED);
-            if (pOverlapped->m_isArray)
-            {
-                pOverlapped->m_userObjectInternal = static_cast<void*>(OBJECTREFToObject(pOverlapped->m_userObject));
-                ArrayBase* pUserObject = (ArrayBase*)OBJECTREFToObject(pOverlapped->m_userObject);
-                Object **ppObj = (Object**)pUserObject->GetDataPtr(TRUE);
-                size_t num = pUserObject->GetNumComponents();
-                for (size_t i = 0; i < num; i ++)
-                {
-                    callback(ppObj + i, (ScanContext *)lp1, GC_CALL_PINNED);
-                }
-            }
-            else
-            {
-                callback(&OBJECTREF_TO_UNCHECKED_OBJECTREF(pOverlapped->m_userObject), (ScanContext *)lp1, GC_CALL_PINNED);
-            }
-        }
-
-        if (pOverlapped->GetAppDomainId() !=  DefaultADID && pOverlapped->GetAppDomainIndex().m_dwIndex == DefaultADID)
-        {
-            OverlappedDataObject::MarkCleanupNeededFromGC();
-        }
+        GCToEEInterface::WalkOverlappedObjectForPromotion(pPinnedObj, (ScanContext *)lp1, callback);
```

Comment:
Is this going to be called even for regular pinned handles? It would be pretty unfortunate to call this for regular pinned handles. We should split this logic so that it is only called for the async pinned handles.

Reply:
looks like it will. will do.

Reply:
This looks potentially difficult. In this point of the code we are looking at an object reference only and have no idea what kind of handle it came from, so we don't know if it's an async pin handle. It seems the only surefire way to know if we are going to need to do additional work on a given object is to inspect its method table.